### ISSUE TITLE
[SCREENREADER]: STEM Standalone - Alert boxes MUST NOT have a role="alert" unless they are announcing high priority information

### ISSUE THREAD

**Issue by 1Copenut** [department-of-veterans-affairs/va.gov-team#11970]
# [508-defect-2](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

**Feedback framework**

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Description

I (Trevor) offered some feedback in [Design Intent Collab Issue #9124](https://github.com/department-of-veterans-affairs/va.gov-team/issues/9124) about two alert boxes needing `role="alert"` assigned so they would announce immediately to screen readers. I should have clarified that those should be added to individual alert boxes, or the Alertbox component amended with a role `prop` so it could be configured as needed.

The VSP Frontend Tools team had to [revert the hardcoded change today](https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/pull/363) with pull 363 because all Alertboxes would read out their content immediately to screen readers. This was confirmed by the VSA team during a staging review for a new feature earlier today.

We should work together to create a new change to the Alertbox component with the new success criteria below.

I'll list the issue and pull request in a related issues section.

/cc @cvalarida && @rianfowler for awareness.

## Point of Contact

**VFS Point of Contact:** _Trevor, Rian_

## Acceptance Criteria

- [ ] Alert boxes should not have a role by default
- [ ] Alert boxes could have a Boolean prop that can be passed to add a role="alert"
- [ ] Alert boxes that have a role="alert" should have the attribute applied to the containing `<div>`
- [ ] Screen readers should interrupt normal document reading to announce Alertboxes with a role="alert"
- [ ] Screen readers should NOT interrupt normal document reading if the Alertbox does not have a role="alert"

## WCAG or Vendor Guidance (optional)

* [Meaningful Sequence: Understanding SC 1.3.2](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-sequence.html)
* [MDN ARIA: alert role](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Alert_Role)

**Comment by shawkey-daniel** [comment]
The code that initiated the initial request has been refactored to not use the AlertBox component. This can be worked in the future, but is not a priority for the BAH team at this time.

**Comment by 1Copenut** [comment]
Thank you @shawkey-daniel for the update.

**Comment by zurbergram** [comment]
https://github.com/department-of-veterans-affairs/vets-website/pull/14270

**Comment by delli-gatti-michael** [comment]
@1Copenut These changes are in staging. I tried a few different browser/screen reader combinations and I don't think we've achieved the result we were going for. My findings are below:

Voiceover & Chrome: Seems to behave as expected where the entire Alert is read as soon as the page loads
Voiceover and Safari: Seems to behave as expected where the entire Alert is read as soon as the page loads

JAWS and Chrome: Reads the entire alert as soon as the page loads
JAWS and IE: Treats the page as if there is no alert
JAWS and Edge: Treats the page as if there is no alert

NVDA & Chrome: Reads the alert as soon as the page loads, but skips the header
NVDA and IE: Treats the page as if there is no alert
NVDA and Edge: Treats the page as if there is no alert
NVDA and Firefox: Treats the page as if there is no alert

I ran an axe check on both pages as well. Looks like the first "Based on your response, you may not be eligible" page has a heading level issue, but I think that's covered in a separate defect.

**Comment by 1Copenut** [comment]
@delli-gatti-michael Thank you for the comprehensive testing. NVDA and JAWs with some browsers might not be honoring the `role="alert"` for a couple of reasons I can think of:

1. The role is applied to a `div[role="alert"]` instead of a single text element like a `<span>` or `<p>`
2. The role is on a `<div>` not on the page on first render.

What happens if you move the `role="alert"` declaration up further on the page, into a div that is added to the DOM with the first `render()` call?

**Comment by zurbergram** [comment]
@1Copenut @delli-gatti-michael 
https://github.com/department-of-veterans-affairs/vets-website/pull/14426

JAWS+EDGE: i moved the `role="alert"` to most parent div of `initial-confirm-eligiblity` and nothing
and to a `p` tag and nothing

**Comment by desireeturner** [comment]
@1Copenut any suggestions on path forward on this issue?

**Comment by 1Copenut** [comment]
@desireeturner I've been struggling with this one. After a lot of research and code experiments, I can say a couple of things for sure:

* Alerts (items with `role="alert"`) that are on the page on first load will not announce in some screen readers, no matter what we do.
* Adding alerts after the page loads is more consistent. These dynamic loads will announce themselves immediately to screen readers. 

I wrote a small Codepen ( https://codepen.io/tpierce_402/pen/GRorBrL ) to demonstrate the difference. There's one alert container on the page all the time, and a second one that is added by pressing a button.

I have a new idea. Instead of trying to get screen readers to read the alert boxes on first load, can we add `tabindex="0"` to both of the alert boxes in the InVision links below. I was talking with some a11y specialists this weekend who mentioned in a separate thread that AT users often navigate forms using the `TAB` key. If we make those alerts focusable, they're much more likely to be read aloud by screen readers because they'll be tab stops now.

* https://bahdigital.invisionapp.com/share/MUIADS5R2HC#/screens/296036932
* https://bahdigital.invisionapp.com/share/MUIADS5R2HC#/screens/296036941

**Comment by desireeturner** [comment]
@delli-gatti-michael @zurbergram see @1Copenut's comments above.

**Comment by zurbergram** [comment]
@1Copenut @delli-gatti-michael 
See https://github.com/department-of-veterans-affairs/vets-website/pull/14426 it has been updated with the `tabindex="0"` changes for those alerts

**Comment by 1Copenut** [comment]
@zurbergram Can we adjust this blue information box about active duty to remove the `role="alert"` and add the `tabindex="0"` like the other two? I'll re-test them all asap. Thank you!

---
![Screen Shot 2020-12-01 at 5.38.32 PM.png](https://images.zenhubusercontent.com/5ac217b74b5806bc2bcd3fc8/5d4325d2-1514-447a-aad9-af32b3f745b5)

**Comment by zurbergram** [comment]
@1Copenut see https://github.com/department-of-veterans-affairs/vets-website/pull/15143

**Comment by 1Copenut** [comment]
@zurbergram & @desireeturner Just retested in Staging. The three alert boxes are all taking keyboard focus as expected. Moving this one to close. Thank you!
